cpp/code/flight.cc

```diff
@@ -0,0 +1,300 @@
+    *stream = std::unique_ptr<arrow::flight::FlightDataStream>(
+        new arrow::flight::RecordBatchStream(owning_reader));
+
+    return arrow::Status::OK();
+  }
+
+  arrow::Status ListActions(
+      const arrow::flight::ServerCallContext&,
+      std::vector<arrow::flight::ActionType>* actions) override {
+    *actions = {kActionDropDataset};
+    return arrow::Status::OK();
+  }
+
+  arrow::Status DoAction(const arrow::flight::ServerCallContext&,
+                         const arrow::flight::Action& action,
+                         std::unique_ptr<arrow::flight::ResultStream>* result) {
+    if (action.type == kActionDropDataset.type) {
+      *result = std::unique_ptr<arrow::flight::ResultStream>(
+          new arrow::flight::SimpleResultStream({}));
+      return DoActionDropDataset(action.body->ToString());
+    }
+    return arrow::Status::NotImplemented("Unknown action type: ", action.type);
+  }
+
+ private:
+  arrow::Result<arrow::flight::FlightInfo> MakeFlightInfo(
+      const arrow::fs::FileInfo& file_info) {
+    ARROW_ASSIGN_OR_RAISE(auto input, root_->OpenInputFile(file_info));
+    std::unique_ptr<parquet::arrow::FileReader> reader;
+    ARROW_RETURN_NOT_OK(parquet::arrow::OpenFile(
+        std::move(input), arrow::default_memory_pool(), &reader));
+
+    std::shared_ptr<arrow::Schema> schema;
+    ARROW_RETURN_NOT_OK(reader->GetSchema(&schema));
+
+    auto descriptor =
+        arrow::flight::FlightDescriptor::Path({file_info.base_name()});
+
+    arrow::flight::FlightEndpoint endpoint;
+    endpoint.ticket.ticket = file_info.base_name();
+    arrow::flight::Location location;
+    ARROW_RETURN_NOT_OK(
+        arrow::flight::Location::ForGrpcTcp("localhost", port(), &location));
+    endpoint.locations.push_back(location);
+
+    int64_t total_records = reader->parquet_reader()->metadata()->num_rows();
+    int64_t total_bytes = file_info.size();
+
+    return arrow::flight::FlightInfo::Make(*schema, descriptor, {endpoint},
+                                           total_records, total_bytes);
+  }
+
+  arrow::Result<arrow::fs::FileInfo> FileInfoFromDescriptor(
+      const arrow::flight::FlightDescriptor& descriptor) {
+    if (descriptor.type != arrow::flight::FlightDescriptor::PATH) {
+      return arrow::Status::Invalid("Must provide PATH-type FlightDescriptor");
+    } else if (descriptor.path.size() != 1) {
+      return arrow::Status::Invalid(
+          "Must provide PATH-type FlightDescriptor with one path component");
+    }
+    return root_->GetFileInfo(descriptor.path[0]);
+  }
+
+  arrow::Status DoActionDropDataset(const std::string& key) {
+    return root_->DeleteFile(key);
+  }
+
+  std::shared_ptr<arrow::fs::FileSystem> root_;
+};  // end ParquetStorageService
+
+TEST(ParquetStorageServiceTest, PutGetDelete) {
+  StartRecipe("ParquetStorageService::StartServer");
+  auto fs = std::make_shared<arrow::fs::LocalFileSystem>();
+  ASSERT_OK(fs->CreateDir("./flight_datasets/"));
```

Comment:
Wouldn't it be better to define a helper function returning a `Status`, and use `ARROW_RETURN_NOT_OK`  / `ARROW_ASSIGN_OR_RAISE` instead of asserts?

Related, though the solution there is different: https://github.com/apache/arrow-cookbook/pull/52

Reply:
Ah, yes, having that helper in #52 would clean things up here. A helper function would work fine right now as well.

Reply:
Updated (finally, sorry for the delay).